node/network/bridge/src/lib.rs

```diff
@@ -282,37 +324,211 @@ fn construct_view(live_heads: &[Hash]) -> View {
-	let messages: Vec<_> = event_producers.into_iter()
-		.map(|producer| producer(update.clone()))
-		.collect();
-
-	ctx.send_messages(messages).await
-}
-
 async fn update_view(
-	peers: &HashMap<PeerId, PeerData>,
-	live_heads: &[Hash],
 	net: &mut impl Network,
+	ctx: &mut impl SubsystemContext<Message = NetworkBridgeMessage>,
+	live_heads: &[Hash],
 	local_view: &mut View,
-) -> SubsystemResult<Option<NetworkBridgeEvent>> {
+	validation_peers: &HashMap<PeerId, PeerData>,
+	collation_peers: &HashMap<PeerId, PeerData>,
+) -> SubsystemResult<()> {
 	let new_view = construct_view(live_heads);
-	if *local_view == new_view { return Ok(None) }
+	if *local_view == new_view { return Ok(())  }
+
 	*local_view = new_view.clone();
 
-	let message = WireMessage::ViewUpdate(new_view.clone()).encode();
+	send_validation_message(
+		net,
+		validation_peers.keys().cloned(),
+		WireMessage::ViewUpdate(new_view.clone()),
+	).await?;
+
+	send_collation_message(
+		net,
+		collation_peers.keys().cloned(),
+		WireMessage::ViewUpdate(new_view.clone()),
+	).await?;
+
+	if let Err(e) = dispatch_validation_event_to_all(
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	if let Err(e) = dispatch_collation_event_to_all(
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	Ok(())
+}
+
+// Handle messages on a specific peer-set. The peer is expected to be connected on that
+// peer-set.
+async fn handle_peer_messages<M>(
+	peer: PeerId,
+	peers: &mut HashMap<PeerId, PeerData>,
+	messages: Vec<WireMessage<M>>,
+	net: &mut impl Network,
+) -> SubsystemResult<Vec<NetworkBridgeEvent<M>>> {
+	let peer_data = match peers.get_mut(&peer) {
+		None => {
+			net.report_peer(peer, UNCONNECTED_PEERSET_COST).await?;
+
+			return Ok(Vec::new());
+		},
+		Some(d) => d,
+	};
+
+	let mut outgoing_messages = Vec::with_capacity(messages.len());
+	for message in messages {
+		outgoing_messages.push(match message {
+			WireMessage::ViewUpdate(new_view) => {
+				if new_view.0.len() > MAX_VIEW_HEADS {
+					net.report_peer(
+						peer.clone(),
+						MALFORMED_VIEW_COST,
+					).await?;
+
+					continue
+				} else if new_view == peer_data.view {
+					continue
+				} else {
+					peer_data.view = new_view;
+
+					NetworkBridgeEvent::PeerViewChange(
+						peer.clone(),
+						peer_data.view.clone(),
+					)
+				}
+			}
+			WireMessage::ProtocolMessage(message) => {
+				NetworkBridgeEvent::PeerMessage(peer.clone(), message)
+			}
+		})
+	}
+
+	Ok(outgoing_messages)
+}
+
+async fn send_validation_message<I>(
+	net: &mut impl Network,
+	peers: I,
+	message: WireMessage<protocol_v1::ValidationProtocol>,
+) -> SubsystemResult<()>
+	where I: IntoIterator<Item=PeerId>, I::IntoIter: ExactSizeIterator
```

Comment:
Personal style preference: it's much easier for me to understand the set of constraints on generic items when each constraint appears on its own line. I get that you're grouping related items, but I'd prefer to just make the code a bit taller. This appears in several places in this PR.